lucille-plugins/lucille-pinecone/src/main/java/com/kmwllc/lucille/pinecone/indexer/PineconeIndexer.java

```diff
@@ -14,39 +22,38 @@
 import io.pinecone.proto.UpsertRequest;
 import io.pinecone.proto.Vector;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
-import java.util.stream.Collectors;
-
 public class PineconeIndexer extends Indexer {
 
   private static final Logger log = LoggerFactory.getLogger(PineconeIndexer.class);
 
   private final PineconeClient client;
   private final String index;
-  private final List<String> namespaces;
-  private final List<String> embeddingFields;
+  private final Map<String, Object> namespaces;
   private final Set<String> metadataFields;
   private final String mode;
   private PineconeConnection connection;
+  private final String defaultEmbeddingField;
 
   public PineconeIndexer(Config config, IndexerMessenger messenger, String metricsPrefix) {
     super(config, messenger, metricsPrefix);
     this.index = config.getString("pinecone.index");
-    this.namespaces = config.getStringList("pinecone.namespaces");
-    this.embeddingFields = config.getStringList("pinecone.embeddingFields");
+    this.namespaces = config.hasPath("pinecone.namespaces") ? config.getConfig("pinecone.namespaces").root().unwrapped() : null;
     this.metadataFields = new HashSet<>(config.getStringList("pinecone.metadataFields"));
     this.mode = config.hasPath("pinecone.mode") ? config.getString("pinecone.mode") : "upsert";
-    PineconeClientConfig configuration = new PineconeClientConfig()
-        .withApiKey(config.getString("pinecone.apiKey"))
-        .withEnvironment(config.getString("pinecone.environment"))
-        .withProjectName(config.getString("pinecone.projectName"))
+    this.defaultEmbeddingField = ConfigUtils.getOrDefault(config, "pinecone.defaultEmbeddingField", null);
+    PineconeClientConfig configuration = new PineconeClientConfig().withApiKey(config.getString("pinecone.apiKey"))
+        .withEnvironment(config.getString("pinecone.environment")).withProjectName(config.getString("pinecone.projectName"))
         .withServerSideTimeoutSec(config.getInt("pinecone.timeout"));
     this.client = new PineconeClient(configuration);
+
+    if (namespaces == null && defaultEmbeddingField == null) {
+      throw new IllegalArgumentException(
+          "either a mapping of namespaces to embedding fields or a defaultEmbeddingField must be provided");
```

Comment:
"at least one of a defaultEmbeddingField or a non-empty namespaces mapping is required"